programs/token/src/instructions/approve.rs

```diff
@@ -28,11 +30,15 @@ pub struct Approve<'a> {
 
 impl Approve<'_> {
     #[inline(always)]
-    pub fn invoke(&self) -> ProgramResult {
-        self.invoke_signed(&[])
+    pub fn invoke(&self, token_program: TokenProgramVariant) -> ProgramResult {
```

Comment:
Did you consider adding the `token_program` as an account to the struct? That way you don't need to statically "choose" the program, it will depend on the token program account.